analyzer/src/main/kotlin/managers/GoMod.kt

```diff
@@ -235,54 +227,68 @@ class GoMod(
     }
 }
 
-private data class Edge(
-    val source: Identifier,
-    val target: Identifier
-)
-
-private fun Collection<Edge>.getNodes(): Set<Identifier> = flatMap { listOf(it.source, it.target) }.toSet()
-
-private fun Collection<Edge>.toPackageReferenceForest(
-    ignoreNode: (Identifier) -> Boolean = { false }
-): SortedSet<PackageReference> {
-    data class Node(
-        val id: Identifier,
-        val outgoingEdges: MutableSet<Identifier> = mutableSetOf(),
-        val incomingEdges: MutableSet<Identifier> = mutableSetOf()
-    )
-
-    val nodes = mutableMapOf<Identifier, Node>()
-    fun addNode(id: Identifier): Node? = if (!ignoreNode(id)) nodes.getOrPut(id) { Node(id) } else null
-
-    forEach { edge ->
-        val source = addNode(edge.source)
-        val target = addNode(edge.target)
-
-        if (source != null && target != null) {
-            source.outgoingEdges += target.id
-            target.incomingEdges += source.id
-        }
+/**
+ * A class to represent a graph with dependencies. This representation is basically an adjacency list implemented by a
+ * map whose keys are package identifiers and whose values are the identifiers of packages these packages depend on.
+ */
+private class Graph(private val nodeMap: MutableMap<Identifier, Set<Identifier>>) {
```

Comment:
Why is `nodeMap` a constructore parameter? Can that parameter be removed?

Reply:
It is used by `subgraph()`.

Reply:
Missed that, thanks. I believe the following would be still good to do - I'll leave it up to you as I also think it's already good enough.

1. Use a public noarg constructor and make the constructor which accepts the map private
2. Add an pre-condition check to that private constructor which verifies that all identifiers of all sets are contained as key in the given map.